pkg/virt-controller/watch/ovm.go

```diff
@@ -0,0 +1,674 @@
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping OfflineVirtualMachine controller.")
+}
+
+func (c *OVMController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *OVMController) Execute() bool {
+	key, quit := c.Queue.Get()
+	if quit {
+		return false
+	}
+	defer c.Queue.Done(key)
+	if err := c.execute(key.(string)); err != nil {
+		log.Log.Reason(err).Infof("re-enqueuing OfflineVirtualMachine %v", key)
+		c.Queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed OfflineVirtualMachine %v", key)
+		c.Queue.Forget(key)
+	}
+	return true
+}
+
+func (c *OVMController) execute(key string) error {
+
+	obj, exists, err := c.vmOVMInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return nil
+	}
+	if !exists {
+		// nothing we need to do. It should always be possible to re-create this type of controller
+		c.expectations.DeleteExpectations(key)
+		return nil
+	}
+	OVM := obj.(*virtv1.OfflineVirtualMachine)
+
+	logger := log.Log.Object(OVM)
+
+	logger.Infof("Working on OVM: %s", OVM.Name)
+
+	//TODO default rs if necessary, the aggregated apiserver will do that in the future
+	if OVM.Spec.Template == nil || len(OVM.Spec.Template.ObjectMeta.Labels) == 0 {
+		logger.Error("Invalid controller spec, will not re-enqueue.")
+		return nil
+	}
+
+	needsSync := c.expectations.SatisfiedExpectations(key)
+
+	ovmKey, err := controller.KeyFunc(OVM)
+	if err != nil {
+		return err
+	}
+
+	// If any adoptions are attempted, we should first recheck for deletion with
+	// an uncached quorum read sometime after listing VirtualMachines (see kubernetes/kubernetes#42639).
+	canAdoptFunc := controller.RecheckDeletionTimestamp(func() (v1.Object, error) {
+		fresh, err := c.clientset.OfflineVirtualMachine(OVM.ObjectMeta.Namespace).Get(OVM.ObjectMeta.Name, &v1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		if fresh.ObjectMeta.UID != OVM.ObjectMeta.UID {
+			return nil, fmt.Errorf("original OfflineVirtualMachine %v/%v is gone: got uid %v, wanted %v", OVM.Namespace, OVM.Name, fresh.UID, OVM.UID)
+		}
+		return fresh, nil
+	})
+	cm := controller.NewVirtualMachineControllerRefManager(controller.RealVirtualMachineControl{Clientset: c.clientset}, OVM, nil, virtv1.OfflineVirtualMachineGroupVersionKind, canAdoptFunc)
+
+	var vm *virtv1.VirtualMachine
+	vmObj, exist, err := c.vmInformer.GetStore().GetByKey(ovmKey)
+	if err != nil {
+		logger.Reason(err).Error("Failed to fetch vm for namespace from cache.")
+		return err
+	}
+	if !exist {
+		logger.Infof("VM not found in cache %s", key)
+		vm = nil
+	} else {
+		vm = vmObj.(*virtv1.VirtualMachine)
+
+		vm, err = cm.ClaimVirtualMachineByName(vm)
+		if err != nil {
+			return err
+		}
+	}
+
+	var createErr error
+
+	// Scale up or down, if all expected creates and deletes were report by the listener
+	// TODO +pkotas update for offlinevirtualmachine
```

Comment:
What is missing here?

Reply:
Nothing is missing. I simply forgot to delete the comment.